compiler/qsc_formatter/src/formatter.rs

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/// This module contains the Q# formatter, which can be used by calling
+/// the format function available from this module. The formatting algorithm
+/// uses the cooked and concrete tokens from the parser crate to create a
+/// token stream of the given source code string. It then uses a sliding window
+/// over this token stream to apply formatting rules when the selected tokens
+/// match certain patterns. Formatting rules will generate text edit objects
+/// when the format of the input string does not match the expected format, and
+/// these edits are returned on using the formatter.
+use qsc_data_structures::span::Span;
+use qsc_frontend::keyword::Keyword;
+use qsc_frontend::lex::{
+    concrete::{self, ConcreteToken, ConcreteTokenKind::*},
+    cooked::TokenKind::*,
+    Delim,
+};
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Debug)]
+pub struct Edit {
+    pub span: Span,
+    pub new_text: String,
+}
+
+impl Edit {
+    fn new(lo: u32, hi: u32, new_text: &str) -> Self {
+        Self {
+            span: Span { lo, hi },
+            new_text: new_text.to_string(),
+        }
+    }
+}
+
+fn make_indent_string(level: usize) -> String {
+    "    ".repeat(level)
+}
+
+/// Applies formatting rules to the given code str, generating edits where
+/// the source code needs to be changed to comply with the format rules.
+pub fn format(code: &str) -> Vec<Edit> {
+    let tokens = concrete::ConcreteTokenIterator::new(code);
+    let mut edits = vec![];
+
+    let mut indent_level = 0;
+
+    // The sliding window used is over three adjacent tokens
+    #[allow(unused_assignments)]
+    let mut one = None;
+    let mut two = None;
+    let mut three = None;
+
+    for token in tokens {
+        // Advance the token window
+        one = two;
+        two = three;
+        three = Some(token);
+
+        let mut edits_for_triple = match (&one, &two, &three) {
+            (Some(one), Some(two), Some(three)) => {
+                // if the token is a {, increase the indent level
+                if let Syntax(Open(Delim::Brace)) = one.kind {
+                    indent_level += 1;
+                }
+
+                // if the token is a }, decrease the indent level
+                if let Syntax(Close(Delim::Brace)) = one.kind {
+                    #[allow(clippy::implicit_saturating_sub)]
+                    if indent_level > 0 {
+                        indent_level -= 1;
+                    }
```

Comment:
This can be written as follows so you don't need the if-branch or the `#[allow(clippy::implicit_saturation_sub)`:
```rust
ident_level = ident_level.saturating_sub(0);
```

Reply:
I had tried to use the `saturation_sub(0)` before, but it didn't seem to produce the right behavior. Maybe I was just not using it correctly. I'll try again...

Reply:
Ah it is because the `indent_level` is a usize, which is ambiguous and can't use the saturating_sub.

Reply:
That is weird, `usize` is actually well defined and has `usize::saturating_sub(rhs)` as a method. It is true that its size depends on the target architecture, but it is still well defined.

Reply:
fixed